src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,219 @@
+using System.Linq;
+using System.Net;
+using System.Threading;
+using DnsClient;
+using DnsClient.Protocol;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Misc;
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    internal enum DnsMonitorState
+    {
+        Created,
+        Running,
+        Failed,
+        Stopped
+    }
+
+    internal class DnsMonitor
+    {
+        #region static
+        private static string EnsureLookupDomainNameIsValid(string lookupDomainName)
+        {
+            Ensure.IsNotNull(lookupDomainName, nameof(lookupDomainName));
+            Ensure.That(lookupDomainName.Count(c => c == '.') >= 2, "LookupDomainName must have at least three components.", nameof(lookupDomainName));
+            return lookupDomainName;
+        }
+
+        private static string GetParentDomainName(string domainName)
+        {
+            var index = domainName.IndexOf('.');
+            return domainName.Substring(index + 1);
+        }
+        #endregion
+
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly IDnsMonitoringCluster _cluster;
+        private readonly LookupClient _lookupClient;
+        private readonly string _lookupDomainName;
+        private readonly string _parentDomainName;
+        private bool _processDnsResultHasEverBeenCalled;
+        private readonly string _query;
+        private DnsMonitorState _state;
+        private Exception _unhandledException;
+
+        private readonly Action<SdamInformationEvent> _sdamInformationEventHandler;
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, string lookupDomainName, IEventSubscriber eventSubscriber, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _lookupDomainName = EnsureLookupDomainNameIsValid(lookupDomainName);
+            _cancellationToken = cancellationToken;
+            _lookupClient = new LookupClient();
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _query = "_mongodb._tcp." + _lookupDomainName;
+            _state = DnsMonitorState.Created;
+
+            eventSubscriber?.TryGetEventHandler(out _sdamInformationEventHandler);
+        }
+
+        // public properties
+        public DnsMonitorState State => _state;
+
+        public Exception UnhandledException => _unhandledException;
+
+        // public methods
+        public void Start()
+        {
+            _state = DnsMonitorState.Running;
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception exception)
+            {
+                _state = DnsMonitorState.Failed;
+                _unhandledException = exception;
+
+                if (_sdamInformationEventHandler != null)
```

Comment:
I guess, this code can be moved to a helper method, but up to you.

Reply:
I'm fine with leaving it as is.